### ISSUE TITLE
Refresh CCMA code

### ISSUE THREAD

**Issue by rmcgibbo** [pandegroup/openmm#1040]
@peastman and I spent some time looking at the CCMA code today, as part of trying to build a new test case. The code is kind of hard to follow mathematically, and doesn't really track exactly with the paper.

- The [scaling](https://github.com/pandegroup/openmm/blob/bdd0267d1abe39b19d7ed2c7b6912bf222aab09d/platforms/reference/src/SimTKReference/ReferenceCCMAAlgorithm.cpp#L6) of the inverse Jacobian by d[i]/d[j] 2 isn't mentioned, but is critical. (The variable that appears to be the inverse jacobian is not actually the inverse jacobian, it's actually the inverse jacobian scaled by d[i]/d[j]). This is a little tricky and is probably an optimization.
- In the `apply()` function, the [construction of the error term looks](https://github.com/pandegroup/openmm/blob/bdd0267d1abe39b19d7ed2c7b6912bf222aab09d/platforms/reference/src/SimTKReference/ReferenceCCMAAlgorithm.cpp#L301) looks like it's a kind of squared relative error, but in the paper it's an absolute error (not relative) and also not squared. This makes it a little tricky to figure out what's going on during the newton-raphson iterations.
- the theory section of the user guide doesn't have any info about constraint algorithms.

The code is overall pretty ancient, I think. It would be nice to refresh it and add some documentation.

**Comment by jchodera** [comment]
+1

Also, would there be a correction applied to the manuscript (if the scaling is indeed critical)?

**Comment by rmcgibbo** [comment]
I think the scaling is just an optimization that absorbs some constants into the variable holding the inverse jacobian, but I'm not sure exactly how it all comes together.
